### PR TITLE
Bond-Based Fragmenter

### PR DESCRIPTION
**Pull request by ivanorman**
Implements a module "BondBased" that takes a MolecularGraph and a parameter nbonds and returns a FragmentedNuclei object. The fragments returned are all unique fragments consisting of a node in MolecularGraph and all nodes within nbonds edges of that node.

### CODE REVIEW

**Review by ryanmrichard — CHANGES_REQUESTED** [review]
I'd also recommend looking at something like [anthracence](https://en.wikipedia.org/wiki/Anthracene) with n == 7, just to make sure rings don't trip your algorithm up.

**Comment by ryanmrichard** [comment]
As a note on the anthracene, suggestion, for unit testing the only thing that really matters for the algorithm is the number of nodes, and the connectivity of the nodes. In other words, for preparing the input, you shouldn't actually need to get the coordinates of anthracene right, just declare the atoms, assign each one to a node, and then ensure you get the connectivity right.

File: tests/cxx/unit_tests/ghostfragment/fragmenting/bond_based.cpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        chemist::FragmentedNuclei fragment_nodes = chemist::FragmentedNuclei(methane.nuclei());
+        for(std::size_t i = 0; i < 5; ++ i) {
+            fragment_nodes.add_fragment({i});
+        }
+
+        nodes_t nodes(fragment_nodes);
+        graph input(nodes, bonds);
+        mod.change_input("nbonds", std::size_t(1));
+        const auto& rv = mod.run_as<my_pt>(input);
+        return_t corr = frag;
```
[the remark this answers is not in the document]
Comment (ivanorman) [reply tests/cxx/unit_tests/ghostfragment/fragmenting/bond_based.cpp:85]: Fixed, now checks to see if potential new frags are subsets or supersets of any existing frags. If not, the frag is added as normal. If it is a superset of any existing frags, the existing frags are removed and the new frag is added. If it is a proper subset, the new frag is not added.

File: tests/cxx/unit_tests/ghostfragment/fragmenting/bond_based.cpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        for(std::size_t i = 0; i < 24; ++i) {
+            fragment_nodes.add_fragment({i});
+        }
+
+        nodes_t nodes(fragment_nodes);
+        graph input(nodes, bonds);
+        mod.change_input("nbonds", std::size_t(7));
+        const auto& rv = mod.run_as<my_pt>(input);
+        return_t corr = frag;
+        // REQUIRE(corr.operator==(rv));
```
Comment (ivanorman) [inline tests/cxx/unit_tests/ghostfragment/fragmenting/bond_based.cpp:278]: Accidentally left this commented out, fixing now.

**Review by ryanmrichard — APPROVED** [review]

**Pull request merged**